[SPARK-4134] [SPARK-7835] [WIP] Dynamic allocation: Tone down kill error messages

andrewor14 opened the pull request:
**[SPARK-4134]:** We should not log `ERROR` or `WARNING` if executors are killed by the user.
**[SPARK-7835]:** Refactor `HeartbeatReceiverSuite` to increase coverage (needed by SPARK-4134)

~60% of this patch is test code.

---

Right now we get a bunch of scary error messages if the user kills executors manually (or uses dynamic allocation to do so):

![screen shot 2015-05-20 at 3 42 38 pm](https://cloud.githubusercontent.com/assets/2133137/7741351/b8b2abf8-ff32-11e4-9e72-6863ff60c240.png)

This patch tones down these error messages, since these are really not errors. Now the output looks like:

![screen shot 2015-05-21 at 1 17 02 pm](https://cloud.githubusercontent.com/assets/2133137/7758103/dc9012f2-ffbb-11e4-9e17-be0771a34055.png)

I tested this on a real YARN cluster with #6301.

Commit:
Tone down scary unrelated error messages on kill

Commit:
Fix deadlock

We should not askWithReplay in a synchronized block. In this
particular case, the CoarseGrainedSchedulerBackend actor tried
to acquire the same lock when replying, leading to a deadlock.

Commit:
Tone down BlockManagerMasterEndpoint message further

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33214/consoleFull) for   PR 6310 at commit [`677df8a`](https://github.com/apache/spark/commit/677df8a819902036b7e03bd09a4ef5a6d3c50f7b).

harishreedharan commented:
Not sure the whole `Logging` trait etc will allow us to hide the logging per class, but if it does you could change the configuration to log only `ERROR` level messages for `ReliableDeliverySupervisor`

harishreedharan commented:
From what I see here: https://github.com/akka/akka/blob/d9db42b75715cb2ca98c84efbdb8666cb046bfa2/akka-remote/src/main/scala/akka/remote/Endpoint.scala ignoring logging messages below `ERROR` seems ok for that class.

SparkQA commented:
[Test build #33214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33214/consoleFull) for   PR 6310 at commit [`677df8a`](https://github.com/apache/spark/commit/677df8a819902036b7e03bd09a4ef5a6d3c50f7b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33214/
Test FAILed.

Commit:
Tone down even more

andrewor14 commented:
@harishreedharan yeah I'm just going to bump it to `ERROR` in the log4j conf. After an offline discussion with @zsxwing we decided that it's safe to do so.

andrewor14 commented:
By the way I updated the screenshot and fixed a heartbeat receiver race condition that was introduced in this patch previously.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Document the heartbeat receiver race condition

andrewor14 commented:
It would be good for @zsxwing and @vanzin to have a look, since this modifies some code both of you have touched in the recent past.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33277/consoleFull) for   PR 6310 at commit [`3bf89d4`](https://github.com/apache/spark/commit/3bf89d4a07f20c830339e01b2d88dc884f15723f).

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -369,24 +369,47 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+
+      executorsPendingToRemove ++= executorIdsToKill
+
+      numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size
```
@vanzin I believe this is functionally equivalent to what you had in the old code (L384). I believe the only difference is that I moved the `askWithReply` out of the synchronized block. Please let me me know if this is not the case.
```
val newTotal = (numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size
  - filteredExecutorIds.size)
...
executorsPendingToRemove ++= filteredExecutorIds
```

Commit:
Grammar

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33276/
Test FAILed.

SparkQA commented:
[Test build #33279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33279/consoleFull) for   PR 6310 at commit [`95b26b6`](https://github.com/apache/spark/commit/95b26b64581b109c3aef48546fdf4a53e01901f5).

Commit:
Do not forget to attach the listener

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33283 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33283/consoleFull) for   PR 6310 at commit [`d5ab486`](https://github.com/apache/spark/commit/d5ab48671917713457a1c9266e879577500f3bfc).

vanzin commented:
Looks sane to me.

SparkQA commented:
[Test build #33279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33279/consoleFull) for   PR 6310 at commit [`95b26b6`](https://github.com/apache/spark/commit/95b26b64581b109c3aef48546fdf4a53e01901f5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33279/
Test PASSed.

SparkQA commented:
[Test build #33283 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33283/consoleFull) for   PR 6310 at commit [`d5ab486`](https://github.com/apache/spark/commit/d5ab48671917713457a1c9266e879577500f3bfc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33283/
Test PASSed.

SparkQA commented:
**[Test build #33277 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33277/consoleFull)**     for PR 6310 at commit [`3bf89d4`](https://github.com/apache/spark/commit/3bf89d4a07f20c830339e01b2d88dc884f15723f)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33277/
Test FAILed.

Commit:
Guard against heartbeat receiver race again

Now the heartbeat receiver will never create a new entry for an
executor that has not registered.

andrewor14 commented:
After testing further I found that the race condition wasn't fully resolved. This should be actually fixed in the latest commit.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33311/consoleFull) for   PR 6310 at commit [`17af2ee`](https://github.com/apache/spark/commit/17af2eea7d13164218eb189c92d41925e9f47bf6).

SparkQA commented:
[Test build #33311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33311/consoleFull) for   PR 6310 at commit [`17af2ee`](https://github.com/apache/spark/commit/17af2eea7d13164218eb189c92d41925e9f47bf6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33311/
Test FAILed.

Commit:
Fix compile + revamp tests for HeartbeatReceiver

This covers not only the change in behavior introduced in this
patch, but also existing behavior (e.g. expire dead hosts) that
was simply not tested. This commit also refactors the test in a
way that eliminates the existing duplicate code.

Commit:
Merge branch 'master' of github.com:apache/spark into da-tone-down

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33366/consoleFull) for   PR 6310 at commit [`ec1cbf5`](https://github.com/apache/spark/commit/ec1cbf5e9f119c910f966d33f9d2d72352cd56f1).

andrewor14 commented on core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:
```diff
@@ -92,24 +102,38 @@ private[spark] class HeartbeatReceiver(sc: SparkContext)
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case ExpireDeadHosts =>
+      expireDeadHosts()
+      context.reply(true)
```
moved this here to keep the tests deterministic

Commit:
HeartbeatReceiverSuite: increase determinism

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33368/consoleFull) for   PR 6310 at commit [`44f0617`](https://github.com/apache/spark/commit/44f0617813c75f95d69096beeaea911a8aca5ba8).

andrewor14 commented on core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:
```diff
@@ -92,24 +102,39 @@ private[spark] class HeartbeatReceiver(sc: SparkContext)
+      context.reply(true)
+    case ExpireDeadHosts =>
+      expireDeadHosts()
+      context.reply(true)
```
I moved these here to increase test determinism

Commit:
Fix tests + unmatched error

It turns out that `send` doesn't actually transmit messages to
`receiveAndReply`. This means we need to duplicate a few messages
in both `receive` and `receiveAndReply` if we want both `send`
and `askWithRetry` to work.

Commit:
Fix `self`

It's a method, not a value.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33374 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33374/consoleFull) for   PR 6310 at commit [`45b134b`](https://github.com/apache/spark/commit/45b134b3ec80e54beb802f70a37d57e74237533d).

SparkQA commented:
[Test build #33374 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33374/consoleFull) for   PR 6310 at commit [`45b134b`](https://github.com/apache/spark/commit/45b134b3ec80e54beb802f70a37d57e74237533d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33374/
Test FAILed.

Commit:
Remove duplicate code by always using ask

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33376 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33376/consoleFull) for   PR 6310 at commit [`9337e9d`](https://github.com/apache/spark/commit/9337e9d9752a70b1e10f8151be30cc8aeabcbe19).

zsxwing commented on core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:
```diff
@@ -43,15 +43,25 @@ private[spark] case class Heartbeat(
+  extends ThreadSafeRpcEndpoint with SparkListener with Logging {
+
+  def this(sc: SparkContext) {
+    this(sc, new SystemClock)
```
Could you make `SystemClock` as a default value to save one constructor, please?

SparkQA commented:
[Test build #33368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33368/consoleFull) for   PR 6310 at commit [`44f0617`](https://github.com/apache/spark/commit/44f0617813c75f95d69096beeaea911a8aca5ba8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33368/
Test FAILed.

SparkQA commented:
[Test build #33366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33366/consoleFull) for   PR 6310 at commit [`ec1cbf5`](https://github.com/apache/spark/commit/ec1cbf5e9f119c910f966d33f9d2d72352cd56f1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33366/
Test PASSed.

SparkQA commented:
[Test build #33376 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33376/consoleFull) for   PR 6310 at commit [`9337e9d`](https://github.com/apache/spark/commit/9337e9d9752a70b1e10f8151be30cc8aeabcbe19).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExecutorLossReason(val message: String, val isError: Boolean = true) `
  * `  case class RemoveExecutor(executorId: String, reason: String, isError: Boolean)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33376/
Test FAILed.

andrewor14 commented:
I discovered a few more issues with this patch. Putting this on hold for now.

andrewor14 commented on core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala:
```diff
@@ -17,65 +17,145 @@
+import org.apache.spark.util.ManualClock
 
-class HeartbeatReceiverSuite extends FunSuite with LocalSparkContext {
+class HeartbeatReceiverSuite
```
I'm going to factor the changes in this class out to a separate patch

andrewor14 commented:
I'm going to open a new patch later once I spend some time to figure out the issue with this one. In the mean time let's not block the refactoring of the `HeartbeatReceiverSuite` on this.
